exe/Main.hs

```diff
@@ -188,22 +192,256 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath' -> file) diags) =
+                          setOptions opts (hsc_dflags hscEnv)
+        dep_info <- getDependencyInfo (componentDependencies opts)
+        -- Now lookup to see whether we are combining with an exisiting HscEnv
+        -- or making a new one. The lookup returns the HscEnv and a list of
+        -- information about other components loaded into the HscEnv
+        -- (unitId, DynFlag, Targets)
+        modifyVar hscEnvs $ \m -> do
+            -- Just deps if there's already an HscEnv
+            -- Nothing is it's the first time we are making an HscEnv
+            let oldDeps = Map.lookup hieYaml m
+            let -- Add the raw information about this component to the list
+                -- We will modify the unitId and DynFlags used for
+                -- compilation but these are the true source of
+                -- information.
+                new_deps = (thisInstalledUnitId df, df, targets, cfp, dep_info) : maybe [] snd oldDeps
+                -- Get all the unit-ids for things in this component
+                inplace = map (\(a, _, _, _, _) -> a) new_deps
+                -- Remove all inplace dependencies from package flags for
+                -- components in this HscEnv
+                rearrange (uid, (df, uids), ts, cfp, di) = (uid, (df, uids, ts, cfp, di))
+                do_one (uid,df, ts, cfp, di) = rearrange (uid, removeInplacePackages inplace df, ts, cfp, di)
+                -- All deps, but without any packages which are also loaded
+                -- into memory
+                new_deps' = map do_one new_deps
+            -- Make a new HscEnv, we have to recompile everything from
+            -- scratch again (for now)
+            -- It's important to keep the same NameCache though for reasons
+            -- that I do not fully understand
+            print ("Making new HscEnv" ++ (show inplace))
+            hscEnv <- case oldDeps of
+                        Nothing -> emptyHscEnv
+                        Just (old_hsc, _) -> setNameCache (hsc_NC old_hsc) <$> emptyHscEnv
+            newHscEnv <-
+              -- Add the options for the current component to the HscEnv
+              evalGhcEnv hscEnv $ do
+                _ <- setSessionDynFlags df
+                getSession
+            -- Modify the map so the hieYaml now maps to the newly created
+            -- HscEnv
+            -- Returns
+            -- * the new HscEnv so it can be used to modify the
+            --   FilePath -> HscEnv map
+            -- * The information for the new component which caused this cache miss
+            -- * The modified information (without -inplace flags) for
+            --   existing packages
+            pure (Map.insert hieYaml (newHscEnv, new_deps) m, (newHscEnv, head new_deps', tail new_deps'))
+
+
+    session <- return $ \(hieYaml, cfp, opts) -> do
+        (hscEnv, new, old_deps) <- packageSetup (hieYaml, cfp, opts)
+        -- TODO Handle the case where there is no hie.yaml
```

Comment:
What happens here atm?

Reply:
There is one session which contains everything which doesn't have a hie.yaml file. This is probably wrong..